src/lexer.rs

```diff
@@ -182,6 +190,65 @@ impl<'a> Iterator for ReplaceComments<'a> {
 // preprocessor: # and newlines. It also include metadata for the token for whether it is at the
 // start of the line, or if it has leading whitespace.
 
+// This is a helper iterator to abstract away the tracking of location data (offset, line) from
+// `Lexer`. It looks like a Peekable<Iterator<char>> with `next_char` and `peek_char` but also
+// allows querying the last seen/consumed lines / offset.
+#[derive(Clone)]
+struct LexerCharIterator<'a> {
+    inner: ReplaceComments<'a>,
+    peeked: Option<(CharAndLine, *const u8)>,
+    last_consumed: (CharAndLine, *const u8),
+    input_start: *const u8,
+}
+
+pub const NONE_CONSUMED_SENTINEL_VALUE: char = '\r';
+
+impl<'a> LexerCharIterator<'a> {
+    pub fn new(input: &'a str) -> Self {
+        LexerCharIterator {
+            inner: ReplaceComments::new(input),
+            peeked: None,
+            last_consumed: ((NONE_CONSUMED_SENTINEL_VALUE, 0), input.as_bytes().as_ptr()),
+            input_start: input.as_bytes().as_ptr(),
+        }
+    }
+    fn next_char(&mut self) -> Option<char> {
+        self.last_consumed = match self.peeked.take() {
+            Some(v) => v,
+            None => {
+                let ptr = self.inner.get_current_ptr();
+                (self.inner.next()?, ptr)
+            }
+        };
+        Some(self.last_consumed.0 .0)
+    }
+
+    fn peek_char(&mut self) -> Option<char> {
+        match self.peeked {
+            Some(v) => Some(v.0 .0),
```

Comment:
nit: This extra space looks really weird can we run `cargo fmt` to fix this

Reply:
It was actually introduced by `cargo fmt` :/

Reply:
Just found out about https://github.com/rust-lang/rustfmt/issues/4779 well then there's nothing we can do then

Reply:
We could add a rustfmt.toml with `version = "Two"` and we should check formatting on CI too

Reply:
If I'm not mistaken that requires a nightly toolchain

Reply:
Just tried it, it's only available in the nightly channel.